[2022.10.21] Setting: Redux 초기 세팅 및 컴포넌트 데이터 연동

0jaaack opened the pull request:
### Redux 초기 세팅 및 컴포넌트 데이터 연동

### Description

- redux 설치 및 초기 설정 완료하였습니다.
- redux-toolkit 적용하였습니다.
- 기존에 전역 상태 관리가 필요했던 상태들(pageSequence, pptData)을, 하나의 slice로 두었습니다.

### Point

- 현재 Footer 컴포넌트에서 이전 sequence로 이동하는 동작을, 두 개의 dispatch를 묶어서 사용하고 있는데요. 따로 액션을 만들어서 처리해야 할 지, 아니면 지금 방식 그대로 처리할 지 고민입니다. 좋은 방안 있으면 논의 부탁드립니다!
- `pageSequence` 관리 방식을, 한 컴포넌트 안에서 switch를 통한 분기처리로 진행하고 있는데요. 더 효율적인 관리 방안이 있을 지 고민입니다.

### ETC

- [상태관리 라이브러리 필요성 및 선택 이슈](https://www.notion.so/vanillacoding/e04afd501eac446291aa1edceff050b1)

0jaaack committed:
Chore: redux 라이브러리 추가 및 Slice 설정/세팅

0jaaack committed:
Setting: redux 세팅 component와 동기화
1. pageSequence 값에 맞게 Main 컴포넌트 내부 조정 작업
2. pageSequence 값에 맞게 Footer button type 조정
3. pageSequences, viewTypes 상수화 적용

Y-SeonJong commented on src/app/store.js:
```diff
@@ -0,0 +1,16 @@
+const reducer = {
+  pageSequence: pageSequenceReducer,
+  pptData: pptDataReducer,
+};
```
reducer 가 두개 이상인데 reducers는 어떤가여?

0jaaack replied:
reduxToolkit의 configureStore 메소드의 config 값이 reducer 프로퍼티를 가져오기 때문에, reducer라고 이름지었습니다...
상단의 선언부는 reducers라고 이름을 지으면 혼동이 있을 것 같아서, 아래와 통일하였고요...!

https://redux-toolkit.js.org/tutorials/quick-start#create-a-redux-store

Donghyeon0908 replied:
저도 reducer가 더 맞아보입니다

Y-SeonJong commented on src/components/Main.js:
```diff
@@ -1,29 +1,58 @@
-import React, { useState } from "react";
+import React, { useEffect } from "react";
```
useEffect 선언 이유가 무엇인가요????

0jaaack replied:
component data와 redux state 연동 간 작성 후 삭제하지 않은 것 같습니다! 삭제하겠습니다~

Y-SeonJong commented on src/components/Main.js:
```diff
@@ -1,29 +1,58 @@
+        </MainContainer>
+      );
+  }
 }
```
switch 문의 사용 👍👍👍👍👍

Donghyeon0908 commented on src/components/Main.js:
```diff
@@ -1,29 +1,58 @@
-import React, { useState } from "react";
+import React, { useEffect } from "react";
 import styled from "styled-components";
+import { useSelector, useDispatch } from "react-redux";
```
`react, styled-components `가 함께 묶여있는건 저희 개행 컨벤션에 맞지않아보이네요~ 비슷한 내용은 생략하겠습니다!

Y-SeonJong commented on src/components/SlideViewer/FileAttachment.js:
```diff
@@ -1,16 +1,20 @@
 import React from "react";
-
+import { useDispatch } from "react-redux";
```
개행 부탁드립니다!

Donghyeon0908 commented on src/components/SlideViewer/SlideViewer.js:
```diff
@@ -1,12 +1,20 @@
 import SlideList from "./SlideList";
 
-function SlideViewer({ pptData }) {
+function SlideViewer({ viewType, onViewerChanged }) {
+  const viewData = useSelector(({ pptData }) => pptData[viewType]);
+
+  if (!Object.keys(viewData).length) {
+    return <FileAttachment fileType={viewType} onFileAdded={onViewerChanged} />;
+  }
+
```
`SlideViewer`컴포넌트 아래에서 `FileAttachment`가 렌더링 되는건 뭔가 어색해보여요 Viewer는 보여주는 의미가 강하다고 생각하는데 뷰랑 파일 업로드 모두 포괄적으로 묶을 수 있는 컴포넌트 네이밍이 좋지않나생각합니다!

0jaaack replied:
넵 저도 `SlideViewer` 하위에 `FileAttachment` 컴포넌트가 있는 것이 어색하다는 생각을 했는데요. `MainSection`과 같은 네이밍으로 변경하겠습니다!

Donghyeon0908 commented on src/components/layout/Footer.js:
```diff
@@ -1,20 +1,64 @@
+    dispatch(changeSequence(PAGE_SEQUENCES.COMPARE_FILES));
+  };
+
+  switch (pageSequence) {
+    case PAGE_SEQUENCES.ORIGINAL_FILE_ADDED:
+      return (
+        <FooterContainer>
+          <FooterButton onClick={handleOriginalFileUndo}>되돌리기</FooterButton>
+        </FooterContainer>
+      );
+    case PAGE_SEQUENCES.COMPARABLE_FILE_ADDED:
+      return (
+        <FooterContainer>
+          <FooterButton onClick={handleComparableFileUndo}>
+            되돌리기
+          </FooterButton>
+          <FooterButton onClick={handleCompare}>비교하기</FooterButton>
+        </FooterContainer>
+      );
+    case PAGE_SEQUENCES.COMPARE_FILES:
+      return (
+        <FooterContainer>
+          <FooterButton onClick={handleCompareUndo}>되돌리기</FooterButton>
+          <FooterButton>병합하기</FooterButton>
+        </FooterContainer>
+      );
+    case PAGE_SEQUENCES.PREVIEW:
+      return (
+        <FooterContainer>
+          <FooterButton>되돌리기</FooterButton>
           <FooterButton>다운로드</FooterButton>
```
`switch `문을 활용하셔서 footer 버튼을 조건부 렌더링하게 하셨군요 👍

Donghyeon0908 commented on src/components/layout/Footer.js:
```diff
@@ -1,20 +1,64 @@
+        </FooterContainer>
+      );
+    default:
+      return <FooterContainer>adsf</FooterContainer>;
```
adsf 지우는게 좋을거같아요

0jaaack replied:
오타가 들어간 것 같습니다...!

Y-SeonJong commented on src/components/layout/Footer.js:
```diff
@@ -1,20 +1,64 @@
+      );
+    default:
+      return <FooterContainer>adsf</FooterContainer>;
+  }
```
각 타입 별에 따른 설정 굿!!!!

근데 adsf는 뭐에요..?????

Y-SeonJong commented on src/config/constants/pageSequences.js:
```diff
@@ -0,0 +1,10 @@
+  PREVIEW: "PAGE_SEQUENCES/PREVIEW",
+  ERROR: "PAGE_SEQUENCES/ERROR",
+};
+
```
명사 _ 동사 의 형태로 내려오다 FILE_NOT_ADDED ...
동사_명사 COMPARE_FILES, 
가 나오니까 어색한 느낌이 드는데 어떻게 생각하시나요??

0jaaack replied:
해당 부분 반영해서 type 수정완료하였습니다!
```
{
  INITIAL_SEQUENCE: "SEQUENCES/INITIAL_SEQUENCE",
  ADDED_ORIGINAL_FILE: "SEQUENCES/ADDED_ORIGINAL_FILE",
  ADDED_COMPARABLE_FILE: "SEQUENCES/ADDED_COMPARABLE_FILE",
  COMPARISION: "SEQUENCES/COMPARISION",
  PREVIEW: "SEQUENCES/PREVIEW",
  ERROR: "SEQUENCES/ERROR",
}
```

Y-SeonJong commented on src/components/Error/ErrorPage.js:
```diff
@@ -2,7 +2,7 @@ import React from "react";
 import { useNavigate } from "react-router-dom";
 import styled from "styled-components";
```
여긴 제가 개행을 안했었네요 죄송합니다 ㅠㅠ

Y-SeonJong reviewed (commented):
재혁님 Redux 초기 세팅 복잡했을 것 같은데 정말 감사합니다 ㅠㅠ 👍👍👍👍

0jaaack committed:
Fix: 시퀀스 전환 액션 함수 추가 및 변수/타입명 수정
1. chanegeNextSequence, changePreviousSequence 메소드 추가
2. pptDataSlice, sequenceSlice -> pptDataReducer, sequenceReducer로 변경
3. sequence type 타입 명 수정
4. Slide Viewer -> MainSection 컴포넌트 명 수정

Y-SeonJong commented on src/app/App.js:
```diff
@@ -1,28 +1,35 @@
-      </Routes>
-      <Footer buttonType={pageType} />
+      {sequence === SEQUENCES.ERROR ? (
+        <ErrorPage />
```
에러 페이지 위치 좋아요~

0jaaack committed:
Fix: axios POST 요청 로직 복구

Y-SeonJong commented on src/components/MainSection/MainSection.js:
```diff
@@ -0,0 +1,29 @@
+import FileAttachment from "./FileAttachment";
+import SlideList from "./SlideList";
+
+function MainSection({ viewType, onViewerChanged }) {
```
onViewerChanged가 안보여요  ㅠㅠ

0jaaack replied:
onViewerChanged는 미사용하는 props인데 삭제를 안했습니다.. 삭제하겠습니다!

Donghyeon0908 reviewed (approved):
재혁님 redux설정에 고생하셨습니다! 중요 로직 이제 전부 작성되었고 기능구현에 힘냅시다!

Y-SeonJong commented on src/config/constants/sequences.js:
```diff
@@ -0,0 +1,18 @@
+  COMPARISION: "SEQUENCES/COMPARISION",
+  PREVIEW: "SEQUENCES/PREVIEW",
+  ERROR: "SEQUENCES/ERROR",
+};
```
👍

Y-SeonJong commented on src/features/sequenceReducer.js:
```diff
@@ -0,0 +1,28 @@
+    return SEQUENCE_FLOW[previousIndex];
+  },
+});
+
```
페이지를 순서별로 나열하는거 나쁘지 않은것 같습니다.

Y-SeonJong reviewed (approved):
이제 시작이다~~~ 정말 수고 많으셨어여 바통터치 시작하겠습니다.

0jaaack committed:
Fix: 불필요 props 삭제
